articles/151_node_to_participant_mapping.md

```diff
@@ -0,0 +1,213 @@
+---
+layout: default
+title: Node to Participant mapping
+permalink: articles/node_to_participant_mapping.html
+abstract: This article analyze the performance implications of having a one-to-one Node to Participant mapping and propose an alternative approach.
+  How Nodes are mapped to DDS participants.
+author: '[Ivan Paunovic](https://github.com/ivanpauno)'
+published: true
+categories: Middleware
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Introduction
+
+### What is a `Node`?
+
+In ROS, a `Node` is an entity used to group other entities.
+For example: `Publishers`, `Subscriptions`, `Services`, `Clients`.
+`Nodes` ease organization and code reusage, as they can be composed or launched in different ways.
+
+### What is a `Domain Participant`?
+
+A `Domain Participant` is a type of DDS entity.
+`Participants` also group other entities, like `Publishers`, `Subscribers`, `Data Writters`, `Data Readers`, etc.
+But participants do more than that:
+
+- Each `Participant` does discovery by its own.
+  Creating more than one `Participant` increase cpu and network.
+- Each `Participant` keeps track of the others and of each DDS entity.
+  Using more than one will duplicate that data.
+- Each `Participant` creates multiple threads for event handling, discovery, etc.
```

Comment:
@ivanpauno I'd bind this statement with particular DDS implementations (as it's an implementation detail).

Reply:
I'm not sure if it's part of the standard or not.
I will rephrase to: `... may create`, and add a cite to `connext` documentation.

Running a listener demo with fastrtps and running `ps -elfT`:
```
0 S ivanpau+ 25754 25754 25749  0  80   0 - 108451 futex_ 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25755 25749  0  80   0 - 108451 futex_ 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25756 25749  0  80   0 - 108451 ep_pol 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25757 25749  0  80   0 - 108451 skb_wa 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25758 25749  0  80   0 - 108451 skb_wa 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25759 25749  0  80   0 - 108451 skb_wa 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25760 25749  0  80   0 - 108451 futex_ 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
```
7 threads! and it's using a single threaded executor (some of them may be from ROS).

Doing the same with `manual_composition` demo creates 19 threads.
The demo compose 4 nodes.
I guess FastRTPS creates 4 nodes per participant, and we always create two background threads independently of the number of Nodes (one is for ctrl-c, the other IDK it may come from fastrtps too).

Composing 20 nodes would create 83 threads.